### ISSUE TITLE
malformed_stable_exception does not provide enough info t

### ISSUE THREAD

**Issue by tzach** [scylladb/scylladb#669]
Error message `WARNING: exceptional future ignored of type 'sstables::malformed_stable_exception’`
does not provide enough info to debug the issue.
Should provide keyspace, table, sstable name.

**Comment by raphaelsc** [comment]
@tzach How did you trigger the exception? I would like to see 'who' is ignoring the exception.

**Comment by tzach** [comment]
@raphaelsc, sorry this is from a user question I do not have more data.

**Comment by benoit-canet** [comment]
I think providing additional context is super difficult to do on most call sites because this exception is thrown from the various following places:

database.cc:                            throw sstables::malformed_sstable_exception("Invalid State encountered. TOC file already processed",
database.cc:                            throw sstables::malformed_sstable_exception("Invalid State encountered. Temporary TOC file found after TOC file was processed",
database.cc:                    throw sstables::malformed_sstable_exception(sprint("At directory: %s: no TOC found for SSTable with generation %d!. Refusing to boot", sstdir, v.first),
sstables/consumer.hh:                throw sstables::malformed_sstable_exception("unknown prestate");
sstables/consumer.hh:                    throw sstables::malformed_sstable_exception(
sstables/exceptions.hh:class malformed_sstable_exception : public std::exception {
sstables/exceptions.hh:    malformed_sstable_exception(sstring msg,
sstables/exceptions.hh:struct bufsize_mismatch_exception : malformed_sstable_exception {
sstables/exceptions.hh:        malformed_sstable_exception(sprint("Buffer improperly sized to hold requested data. Got: %ld. Expected: %ld", size, expected))
sstables/index_reader.hh:            throw malformed_sstable_exception("unknown state");
sstables/partition.cc:            throw malformed_sstable_exception(sprint("Found %d clustering elements in column name. Was not expecting that!", clustering.size()));
sstables/partition.cc:                        throw malformed_sstable_exception("Static row has clustering key information. I didn't expect that!");
sstables/partition.cc:                throw malformed_sstable_exception(sprint("schema does not contain column: %s", cell.c_str()));
sstables/partition.cc:                throw malformed_sstable_exception("frozen set should behave like a cell\n");
sstables/partition.cc:            throw malformed_sstable_exception(sprint("Key mismatch. Got %s while processing %s", to_hex(bytes_view(key)).c_str(), to_hex(bytes_view(_key)).c_str()));
sstables/row.cc:                throw malformed_sstable_exception("FIXME COUNTER_MASK");
sstables/row.cc:                    throw malformed_sstable_exception("FIXME COUNTER_UPDATE_MASK");
sstables/row.cc:                        throw malformed_sstable_exception("deleted cell expects local_deletion_time value");
sstables/row.cc:                    throw malformed_sstable_exception("deleted cell expects local_deletion_time value");
sstables/row.cc:            throw malformed_sstable_exception("unknown state");
sstables/row.cc:            throw malformed_sstable_exception("end of input, but not end of row");
sstables/sstables.cc:            throw malformed_sstable_exception("Estimated histogram with zero size found. Can't continue!");
sstables/sstables.cc:                throw malformed_sstable_exception("SSTable too big: " + to_sstring(size) + " bytes.");
sstables/sstables.cc:                    throw malformed_sstable_exception("Unrecognized TOC component: " + c);
sstables/sstables.cc:                throw malformed_sstable_exception("Empty TOC");
sstables/sstables.cc:                throw malformed_sstable_exception(file_path + ": file not found");
sstables/sstables.cc:                throw malformed_sstable_exception(file_path + ": file not found");
sstables/sstables.cc:        throw malformed_sstable_exception("Current sampling level (" + to_sstring(downsampling::BASE_SAMPLING_LEVEL) + ") not enough to generate summary.");
sstables/sstables.cc:        throw malformed_sstable_exception(sprint("invalid version for file %s. Name doesn't match any known version.", fname));
tests/sstable_test.cc:        } catch (malformed_sstable_exception& e) {

Some don't have any relevant informations concerning the context.

Maybe we should fix this bug by properly reproducing it by fuzzing SSTables writes with CharybdeFS.

@slivne : What do you think about this idea ?

**Comment by benoit-canet** [comment]
I will start by writing a patch to get the exception build site file and line number
so we will be able to catch it properly at next occurence.

**Comment by avikivity** [comment]
We have a few entry points (sstable::read_range_rows, or rather sstable::mutation_reader::impl::read()); they can catch a m_s_e, add the sstable name, and rethrow it.

**Comment by avikivity** [comment]
@benoit-canet instead of file/line, the exception string should be properly distinguished.

**Comment by benoit-canet** [comment]
Posted a candidate patch.

**Comment by avikivity** [comment]
Fix does not address the issue.

**Comment by benoit-canet** [comment]
Oh I didn't saw this reply yesterdays I will try that

**Comment by slivne** [comment]
@benoit-canet any update on this ?

**Comment by nyh** [comment]
There are three problems in the message shown on the original issue submission:
1. It doesn't say *which* sstable is malformed.
2. It doesn't say which kind of malformation was found in this file.
3. It doesn't have any context what part of the code forgot to handle this exception.

Regarding 1:
In d5c1366e855ff58cce1da01935b70d2bb1ab0f1f, Glauber added a message to which sstable is malformed, but he only did this for sstables used for compaction. In my review of his patch my comment was that this should have been more general, and not only for compaction.

Regarding 2:
Seastar prints these "exceptional future ignored" messages using report_exception(). For completely general exceptions, this only prints the type's name which is what the original user saw. For classes inheriting from std::exception, it's supposed to print this class's what(), which would show much more information about what went wrong. I don't know why this didn't happen... There is no place in the code which generates a malformed_sstable_exception without a message string. So this needs to be further debugged.

Regarding 3:
This is a more general Seastar issue, which makes debugging these ignored exceptions difficult. But it's still better to see them than not seeing them :-)

All 3 of these problems appear to be open (with the 3rd being most complex and least urgent, and probably shouldn't keep this issue open).

**Comment by slivne** [comment]
@nyh - thanks

@benoit-canet - can you take it from here and address at least 1,2

**Comment by tzach** [comment]
@benoit-canet any updates?

**Comment by benoit-canet** [comment]
@tzach: The direct way of passing the sstable object everywhere seems horrible to do. I have a patch on the list which take another method. @avikivity Do I iterate this patch ? Or do you have a third method ?

**Comment by avikivity** [comment]
Please bump the patch on the list.

**Comment by nyh** [comment]
Another example of this problem: we saw messages like this in the log:
```
storage_proxy - Exception when communicating with 1.1.1.1: RANGE DELETE not implemented. Tried to merge but found gap between <REDACTED> and <REDACTED>.
```
This is an sstable problem, but we don't get to know which sstable had this problem... (by the way, in this specific case it might have been useful also to know the partition key!).
This message comes from a malformed_sstable_exception thrown in partition.cc.

Can we fix this like @glommer fixed the compaction case for #1123 in commit d5c1366 ?

**Comment by tzach** [comment]
@tgrabiec can we close this issue?
Looks like your patch fix the issue

**Comment by tgrabiec** [comment]
@tzach It was actually Pekka's patch. I think it doesn't fix all the places, so this issue should not be closed yet,

**Comment by penberg** [comment]
Exactly. There's still lots of work to do, I only did the easy ones.

**Comment by kostja** [comment]
@tzach this is a bag of bugs. First of all, malformed_sstable_exception does accept a file name and this file name can and is used in a few places. So, formally, this is fixed. However, the bigger issue is that Scylla throws malformed_sstable_exception in many places, where it could have been using more detailed/specialized exceptions, and in some of these places the filename context is missing. Perhaps a different exception altogether could be used for these places.

@bhalevy JFYI

I am closing this now - please feel free to open separate issues for more specific bugs.

**Comment by denesb** [comment]
Closing, we now attach the filename, in addition to the exception message.
